### PR TITLE
Add Delete method to cache.Cache interface

### PR DESCRIPTION
**Pull request by lamida**
**What this PR does**:

- Refactor and simplify cache interfaces in dskit. We are adding the Delete method in the cache.Cache interface so that it can cover the use case on where cache items can be deleted.
- Implement backward compatible cache metrics exposition by moving backend type (e.g. redis/memcached) as label instead of part of metric name. Was part of #260

More context: ([internal link](https://github.com/grafana/backend-enterprise/pull/4030#issuecomment-1379944449)).

Dependency to be updated (will do once this PR merged):
* Backend Enterprise: https://github.com/grafana/backend-enterprise/pull/4113
* Mimir: https://github.com/grafana/mimir/pull/4078

**Which issue(s) this PR fixes**:

Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: cache/cache.go
Comment (pracucci) [inline cache/cache.go:50]: Do we need this interface? Can't we just add `Delete()` to the `Cache` interface?

**Review by pracucci — APPROVED** [review]
LGTM (modulo a comment on the interface). This means we'll implicitly add Redis support to Mimir, which is OK. After this PR will be merged, can you open a PR to Mimir to update dskit?

File: cache/cache.go
Comment (pstibrany) [reply cache/cache.go:50]: While I agree, I'd point out that non-compatible API changes are quite painful for updates.

Comment (pracucci) [reply cache/cache.go:50]: Assuming this interface is used with dskit cache client implementations, is it really an issue?

Comment (pstibrany) [reply cache/cache.go:50]: Yes, changing `Cache` interface will require a cascading update of multiple dependencies in order to update our internal enterprise product. It is doable (we've done it many times, last time earlier this week), just pointing out the pain.

Comment (pracucci) [reply cache/cache.go:50]: I've seen you introduced `CreateClientWithDelete()` in the last commit. Is that necessary?

Comment (56quarters) [reply cache/cache.go:50]: I agree that we should just add `Delete` to the `Cache` interface. It will require cascading changes but I think they'll be straightforward to fix: the dependent projects will either compile or not, no subtle errors to debug.

Comment (lamida) [reply cache/cache.go:50]: > I've seen you introduced CreateClientWithDelete() in the last commit. Is that necessary?
It was intended to create DeletableCache without affecting existing Cache interface useage.

Nevertheless I will just update the implementation by adding Delete method to the cache interface.

File: cache/compression.go
Comment (Logiraptor) [inline cache/compression.go:103]: Given most of these are wrappers around another Cache implementation, I think it would be better to at least delegate the deletion through to the underlying cache. It would be confusing to use snappyCache for example and suddenly deletes stop working.

Comment (lamida) [reply cache/compression.go]: Fixed.

File: cache/mock.go
Comment (pracucci) [inline cache/mock.go:116]: Should call the wrapped one.

File: cache/remote_cache.go
Comment (pracucci) [inline cache/remote_cache.go:96]: Should use the input context.

File: cache/tracing.go
Comment (pracucci) [inline cache/tracing.go:51]: Should call the wrapped one.

File: cache/versioned.go
Comment (pracucci) [inline cache/versioned.go:54]: Why not implemented here?

File: cache/lru.go
Comment (pracucci) [inline cache/lru.go:140]: We should implement it here.

**Review by pracucci — COMMENTED** [review]
Jon, I think you should implement `Delete()` in the various caches. I left comments.

File: cache/versioned.go
Comment (lamida) [reply cache/versioned.go]: Added

File: cache/lru.go
Comment (lamida) [reply cache/lru.go:141]: Added

Comment (Logiraptor) [reply cache/lru.go:141]: I think this also needs to call delete on the underlying cache `l.c.Delete`, right?

Comment (lamida) [reply cache/lru.go]: That's true :palm-face:. Added.

File: cache/mock.go
Comment (56quarters) [inline cache/mock.go:114]: Maybe add a counter for deletes here?

**Review by 56quarters — APPROVED** [review]
LGTM with one suggestion.

**Review by Logiraptor — APPROVED** [review]
LGTM modulo CI passing

**Comment by lamida** [comment]
Please help do one (hopefully) final review @pracucci . I will update backend-enterprise and mimir dependency after this merged. I replied to the comment about context.Background and for now will stop cascading the context from the caller. Let me know what do you think.

File: cache/mock.go
Comment (lamida) [reply cache/mock.go:115]: Added

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:82]:
```suggestion
* [ENHANCEMENT] Cache: Add backward compatible metrics exposition for cache. #255
```

File: cache/lru.go
Comment (pracucci) [inline cache/lru.go:139]: You need to take the `mtx` while accessing the `lru`. Release it before calling `l.c.Delete()`.

File: cache/lru_test.go
Comment (pracucci) [inline cache/lru_test.go:60]: [nit] Please rename the test to `TestLRUCache_StoreFetchDelete`. Please also add an assertion on metrics at the end of this test. It should be something like this:

```
	require.NoError(t, testutil.GatherAndCompare(reg, strings.NewReader(`
		# HELP cache_memory_items_count Total number of items currently in the in-memory cache.
		# TYPE cache_memory_items_count gauge
		cache_memory_items_count{name="test"} 2
		# HELP cache_memory_hits_total Total number of requests to the in-memory cache that were a hit.
		# TYPE cache_memory_hits_total counter
		cache_memory_hits_total{name="test"} 2
		# HELP cache_memory_requests_total Total number of requests to the in-memory cache.
		# TYPE cache_memory_requests_total counter
		cache_memory_requests_total{name="test"} 4
	`)))
```

File: cache/redis_client.go
Comment (pracucci) [inline cache/redis_client.go:95]: This change is fine (actually wanted to keep consistency with the memcached client) but please mention it in the CHANGELOG.

File: cache/remote_cache.go
Comment (pracucci) [inline cache/remote_cache.go:22]: [nit] I would clarify it. Especially `labelName` is very generic.

```suggestion
	labelCacheName             = "name"
	labelBackendName          = "backend"
```

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:82]: Clarify the impact of this change please.

File: cache/versioned_test.go
Comment (pracucci) [inline cache/versioned_test.go:45]: I think here you want to test that V1 doesn't delete V2 too. So I think it's more interesting calling `v2.Fetch()` and ensuring it still returns a value. Then you can try to delete V2 too.

File: cache/remote_cache.go
Comment (pracucci) [inline cache/remote_cache.go:25]: [nit] I would clarify which prefix it is.

```suggestion
	cacheMetricNamePrefix           = "cache_"
```

File: cache/remote_cache.go
Comment (pracucci) [inline cache/remote_cache.go:26]: Pay attention that the "m" should be lowercase. I would also clarify which prefix it is.

```suggestion
	getMultiMetricNamePrefix        = "getmulti_"
```

File: cache/memcached_client.go
Comment (pracucci) [inline cache/memcached_client.go:182]: I would do the naming the other way round. I would call `newRegisterer` just `reg` and the other `legacyReg` to clarify which one is the legacy one.

File: cache/memcached_client.go
Comment (pracucci) [inline cache/memcached_client.go:206]: I don't think this is correct. The final metric name will start with `getmulti_` while it should be `(cache|memcached)_getmulti_*`.

File: cache/memcached_client.go
Comment (pracucci) [inline cache/memcached_client.go:214]: Here you should use `backwardCompatibleRegs` otherwise it's a breaking change.

File: cache/redis_client.go
Comment (pracucci) [inline cache/redis_client.go:180]: [nit] No need for this `if`. The wrappers can handle a nil registry.

File: cache/redis_client.go
Comment (pracucci) [inline cache/redis_client.go:196]: Same problem here. The final one will be `getmulti_cache_*` instead of `cache_getmulti_*`.

File: cache/redis_client.go
Comment (pracucci) [inline cache/redis_client.go:202]: This metric name is not consistent with the memcached client.

**Review by pracucci — COMMENTED** [review]
Thanks Jon for working on this. I left some comments. Generally speaking I don't like very much bundling together two unrelated changes: adding `Delete()` and changing metrics. I think it adds confusion and makes the review harder. Now, given you've already done this work let's keep it bundled this time, but next time they should be two different PRs.

File: cache/remote_cache.go
Comment (lamida) [reply cache/remote_cache.go:26]: Thank you so much for catching this. I did multiple manual tests and this was missed from my visual observation. I will see if I can test to catch this.

File: cache/memcached_client.go
Comment (lamida) [reply cache/memcached_client.go:206]: In my test, it gives me the correct result. Examples:
```
versioned_client_cache_getmulti_gate_queries_concurrent_max{backend="redis",name="admin-cache"} 100
thanos_cache_getmulti_gate_queries_in_flight{backend="memcached",component="query-frontend",name="frontend-cache"} 0
```

So from what I can see, `prometheus.WrapRegisterWithPrefix` will prepend passed registery and add prefix for subsequent wrap or for metrics which will use the Registerer.

I am building a simple program to verify:
```
package main

import (
	"net/http"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	opsProcessed prometheus.Counter
)

func main() {
	reg := prometheus.NewRegistry()
	reg1 := prometheus.WrapRegistererWithPrefix("one_", reg)
	reg2 := prometheus.WrapRegistererWithPrefix("two_", reg1)
	reg3 := prometheus.WrapRegistererWithPrefix("three_", reg2)

	opsProcessed = promauto.With(reg3).NewCounter(prometheus.CounterOpts{
		Name: "myapp_processed_ops_total",
		Help: "The total number of processed events",
	})
	go func() {
		for {
			opsProcessed.Inc()
			time.Sleep(2 * time.Second)
		}
	}()
	http.Handle("/metrics2", promhttp.HandlerFor(reg, promhttp.HandlerOpts{Registry: reg3}))
	http.ListenAndServe(":2112", nil)
}
```

It gives me the following metrics:
```
# HELP one_two_three_myapp_processed_ops_total The total number of processed events
# TYPE one_two_three_myapp_processed_ops_total counter
one_two_three_myapp_processed_ops_total 2
# HELP one_two_three_promhttp_metric_handler_errors_total Total number of internal errors encountered by the promhttp metric handler.
# TYPE one_two_three_promhttp_metric_handler_errors_total counter
one_two_three_promhttp_metric_handler_errors_total{cause="encoding"} 0
one_two_three_promhttp_metric_handler_errors_total{cause="gathering"} 0
```

Please let me know if I might still misunderstand something. :)

File: cache/redis_client.go
Comment (lamida) [reply cache/redis_client.go:194]: Same reply on above https://github.com/grafana/dskit/pull/255#discussion_r1098156614

File: cache/memcached_client.go
Comment (lamida) [reply cache/memcached_client.go]: Actually, I am not adding `backend="memcached"` label intentionally to keep this metric in the old format because `client_info` labels are really client specific and removing memcached from metric name to label might make it confusing. Both redis and memcached will have different set of labels. In this line, `reg` (now renamed to `legacyReg` already has `memcached_` prefix, hence this line will produce the following as in the old format:

```
thanos_memcached_client_info{component="query-frontend",dns_provider_update_interval="30s",max_async_buffer_size="25000",max_async_concurrency="50",max_get_multi_batch_size="100",max_get_multi_concurrency="100",max_idle_connections="100",max_item_size="1048576",name="frontend-cache",timeout="500ms"} 1
versioned_client_memcached_client_info{dns_provider_update_interval="30s",max_async_buffer_size="25000",max_async_concurrency="50",max_get_multi_batch_size="100",max_get_multi_concurrency="100",max_idle_connections="100",max_item_size="1048576",name="admin-cache",timeout="200ms"} 1
```
For redis, I want to have the following format (in this commit still inconsistently produce `versioned_client_cache_redis_client_info`):
```
versioned_client_redis_client_info{backend="redis",connection_pool_size="100",dial_timeout="5s",max_async_buffer_size="25000",max_async_concurrency="50",max_get_multi_batch_size="100",max_get_multi_concurrency="100",max_item_size="1048576",name="admin-cache",read_timeout="3s",write_timeout="3s"} 1
```

Definitely we will maintain backward compatible change for memcached client_info, but do you think we still need to remove cache from metric name and move it to label?

File: cache/redis_client.go
Comment (lamida) [reply cache/redis_client.go:200]: As mentioned in https://github.com/grafana/dskit/pull/255#discussion_r1098161871 memcached will produce `<prefix>_memcached_client_info` as this redis format. The reason to keep cache name in the metric name because different cache will expose different label.

Same question do you think we still need to remove cache from metric name and move it to `backend` label with the above reason?

File: cache/redis_client.go
Comment (lamida) [reply cache/redis_client.go:95]: The first time this is added, the config name and variable name missing max prefix. Ok I will add this new config name in the dskit changelog.

Comment (lamida) [reply cache/redis_client.go:95]: Also, the renaming above is to make the config name consistent with Memcached.

File: CHANGELOG.md
Comment (lamida) [reply CHANGELOG.md]: Changelog is updated with list of metrics with new format.

**Comment by lamida** [comment]
Hi @pracucci thank you for the review yesterday. Yes this PR contains merge from #260. Sorry to make this PR bigger and contain more than one topic.

I have applied all of your PR suggestions and replied to some comments. The main points to summarise here again:
* `prometheus.WrapRegistererWithPrefix` behaviour is replied here: https://github.com/grafana/dskit/pull/255#discussion_r1098156614
* I reply a comment on `client_info` metric name (https://github.com/grafana/dskit/pull/255#discussion_r1098161871) asking whether we can have cache name in the metric name. Nevertheless I still followed your suggestion. But let me know if you think my reply there is reasonable.

File: cache/memcached_client.go
Comment (pracucci) [reply cache/memcached_client.go]: You're right. What you did is correct. Thanks!

File: cache/redis_client.go
Comment (pracucci) [reply cache/redis_client.go]: What you did is correct. I would just add a comment both here and in the memcached one to explain why the metric names here are specific.

File: cache/redis_client.go
Comment (pracucci) [inline cache/redis_client.go:200]: You were convincing about using per-backend specific metric names. Now we're using a common metric name. What has changed?

File: cache/memcached_client.go
Comment (pracucci) [inline cache/memcached_client.go:185]: Ops. Some renaming was applied to the input param too. The input registerer is not legacy, so should just be named `reg`.

I'm suggesting something like this:

```suggestion
	reg prometheus.Registerer,
	name string,
) (*memcachedClient, error) {
	legacyRegister := prometheus.WrapRegistererWithPrefix(legacyMemcachedPrefix, reg)
	reg = prometheus.WrapRegistererWith(
		prometheus.Labels{labelCacheBackend: backendMemcached},
		prometheus.WrapRegistererWithPrefix(cacheMetricNamePrefix, reg))
```

File: cache/redis_client.go
Comment (lamida) [reply cache/redis_client.go:200]: Yes nothing changes. Initially I wanted to keep cache name in the metric name for client_info but now decided to just use backend label as everything else (backward compatible behaviour is kept for memcached_client_info). I just moved "client_info" to common variable name because it is used in two places now.

File: cache/memcached_client.go
Comment (lamida) [reply cache/memcached_client.go:185]: I will check and apply the suggestion once I got back in front of my laptop.

File: cache/redis_client.go
Comment (pracucci) [reply cache/redis_client.go:200]: Ok. Sounds good. Don't have a big preference between the two options.

**Review by pracucci — APPROVED** [review]
Thanks Jon for your patience! LGTM modulo [this comment](https://github.com/grafana/dskit/pull/255#discussion_r1098426160).

**Pull request merged**
